plugin-parser/mlops-plugin-parser.cc

```diff
@@ -88,10 +88,128 @@ typedef enum {
+ * @brief Internal enumeration for package manager event types.
+ */
+typedef enum {
+  MLSVC_PKGMGR_MDPARSER_PLUGIN_EVENT_TYPE_INSTALL = 0,
+  MLSVC_PKGMGR_MDPARSER_PLUGIN_EVENT_TYPE_UNINSTALL = 1,
+  MLSVC_PKGMGR_MDPARSER_PLUGIN_EVENT_TYPE_UPGRADE = 2,
+  MLSVC_PKGMGR_MDPARSER_PLUGIN_EVENT_TYPE_MAX
+} mlsvc_package_manager_event_type_e;
+
+/**
+ * @brief Internal function for uninstall models and pipelines.
  */
 static void
-_parse_json (JsonNode *node, mlsvc_json_type_e json_type, const gchar *app_info)
+_uninstall_rpk (const char *name, char *_info, mlsvc_json_type_e json_type)
+{
+  g_autoptr (JsonParser) parser = NULL;
+  g_autoptr (GError) err = NULL;
+  JsonNode *root;
+  JsonObject *object;
+  JsonArray *array = NULL;
+
+  parser = json_parser_new ();
+  guint _info_len = 1U;
+
+  if (!json_parser_load_from_data (parser, _info, -1, &err)) {
+    _E ("Failed to load '%s' info data.", name);
+    return;
+  }
+
+  root = json_parser_get_root (parser);
+  if (!root) {
+    _E ("Failed to load '%s' info, cannot get the top node from info.", name);
+    return;
+  }
+
+  if (JSON_NODE_HOLDS_ARRAY (root)) {
+    array = json_node_get_array (root);
+    if (!array) {
+      _E ("Failed to get root array from '%s' info", name);
+      return;
+    }
+
+    _info_len = json_array_get_length (array);
+  }
+
+  /* Update ML service database. */
+  for (guint i = 0; i < _info_len; ++i) {
+    int ret = 0;
+    if (array)
+      object = json_array_get_object_element (array, i);
+    else
+      object = json_node_get_object (root);
+
+    const gchar *app_info = json_object_get_string_member (object, "app_info");
+    g_autoptr (JsonParser) app_info_parser = json_parser_new ();
+
+    if (!json_parser_load_from_data (app_info_parser, app_info, -1, &err)) {
+      _E ("Failed to load '%s' app_info data.", name);
+      return;
+    }
+
+    JsonNode *app_info_root = json_parser_get_root (app_info_parser);
+    if (!app_info_root) {
+      _E ("Failed to load '%s'app_info, cannot get the top node from app_info.", name);
+      return;
+    }
+
+    JsonObject *app_info_object = json_node_get_object (app_info_root);
+    const gchar *is_rpk = json_object_get_string_member (app_info_object, "is_rpk");
+
+    if (!is_rpk) {
+      _E ("Failed to get 'is_rpk' from '%s' app_info.", name);
+      return;
+    }
+    if (g_ascii_strcasecmp (is_rpk, "F") == 0)
+      continue;
+
+    switch (json_type) {
+      case MLSVC_JSON_MODEL:
+        {
+          const gchar *version = json_object_get_string_member (object, "version");
+          ret = ml_agent_model_delete_force (
+              name, (const uint32_t) g_ascii_strtoull (version, NULL, 10), TRUE);
+
+          if (ret == 0) {
+            _I ("The model is deleted. - name: %s, version %s", name, version);
+          } else {
+            _E ("Failed to delete model return %d. - name: %s, version: %s",
+                ret, name, version);
+          }
+          return;
```

Comment:
check the for loop (array) at line 137
line 181, 193, 198 : this switch-case handles 1st item if given json string is array.

Reply:
thanks, i update it.